server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportShardCloseAction.java

```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Override
+    protected ReplicationResponse newResponseInstance() {
+        return new ReplicationResponse();
+    }
+
+    @Override
+    protected void acquirePrimaryOperationPermit(final IndexShard primary,
+                                                 final ShardCloseRequest request,
+                                                 final ActionListener<Releasable> onAcquired) {
+        primary.acquireAllPrimaryOperationsPermits(onAcquired, request.timeout());
+    }
+
+    @Override
+    protected void acquireReplicaOperationPermit(final IndexShard replica,
+                                                 final ShardCloseRequest request,
+                                                 final ActionListener<Releasable> onAcquired,
+                                                 final long primaryTerm,
+                                                 final long globalCheckpoint,
+                                                 final long maxSeqNoOfUpdateOrDeletes) {
+        replica.acquireAllReplicaOperationsPermits(primaryTerm, globalCheckpoint, maxSeqNoOfUpdateOrDeletes, onAcquired, request.timeout());
+    }
+
+    @Override
+    protected PrimaryResult<ShardCloseRequest, ReplicationResponse> shardOperationOnPrimary(final ShardCloseRequest shardRequest,
+                                                                                            final IndexShard primary) throws Exception {
+        executeShardOperation(primary);
+        return new PrimaryResult<>(shardRequest, new ReplicationResponse());
+    }
+
+    @Override
+    protected ReplicaResult shardOperationOnReplica(final ShardCloseRequest shardRequest, final IndexShard replica) throws Exception {
+        executeShardOperation(replica);
+        return new ReplicaResult();
+    }
+
+    private void executeShardOperation(final IndexShard indexShard) {
+        final ShardId shardId = indexShard.shardId();
+        if (indexShard.getActiveOperationsCount() != 0) {
+            throw new IllegalStateException("On-going operations in progress while checking index shard " + shardId + " before closing");
+        }
+
+        final ClusterBlocks clusterBlocks = clusterService.state().blocks();
+        if (clusterBlocks.hasIndexBlock(shardId.getIndexName(), EXPECTED_BLOCK) == false) {
+            throw new IllegalStateException("Index shard " + shardId + " must be blocked by " + EXPECTED_BLOCK + " before closing");
+        }
+
+        final long maxSeqNo = indexShard.seqNoStats().getMaxSeqNo();
+        if (indexShard.getGlobalCheckpoint() != maxSeqNo) {
+            throw new IllegalStateException("Global checkpoint [" + indexShard.getGlobalCheckpoint()
+                + "] mismatches maximum sequence number [" + maxSeqNo + "] on index shard " + shardId);
+        }
+        indexShard.flush(new FlushRequest());
+        logger.debug("shard {} is ready for closing", shardId);
```

Comment:
can we put the shardId first in the log line so it will look like `[index][0] shard is ready for closing` ?

Reply:
Ok